packages/build/src/service-worker-workbox.ts

```diff
@@ -0,0 +1,173 @@
+
+import {DepsIndex} from './analyzer';
+import {LocalFsPath, posixifyPath} from './path-transformers';
+import {PolymerProject} from './polymer-project';
+
+const logger = logging.getLogger('polymer-build.service-worker');
+
+export interface AddWorkboxServiceWorkerOptions {
+  project: PolymerProject;
+  buildRoot: LocalFsPath;
+  bundled?: boolean;
+  path?: LocalFsPath;
+  workboxConfig?: WorkboxConfig|null;
+  basePath?: LocalFsPath;
+}
+
+/**
+ * Returns an array of file paths for the service worker to precache, based on
+ * the information provided in the DepsIndex object.
+ */
+function getPrecachedAssets(
+    depsIndex: DepsIndex, project: PolymerProject): string[] {
+  const precachedAssets = new Set<string>(project.config.allFragments);
+  precachedAssets.add(project.config.entrypoint);
+
+  for (const depImports of depsIndex.fragmentToFullDeps.values()) {
+    depImports.imports.forEach((s) => precachedAssets.add(s));
+    depImports.scripts.forEach((s) => precachedAssets.add(s));
+    depImports.styles.forEach((s) => precachedAssets.add(s));
+  }
+
+  return Array.from(precachedAssets);
+}
+
+/**
+ * Returns an array of file paths for the service worker to precache for a
+ * BUNDLED build, based on the information provided in the DepsIndex object.
+ */
+function getBundledPrecachedAssets(project: PolymerProject) {
+  const precachedAssets = new Set<string>(project.config.allFragments);
+  precachedAssets.add(project.config.entrypoint);
+
+  return Array.from(precachedAssets);
+}
+
+// Matches URLs like "/foo.png/bar" but not "/foo/bar.png".
+export const hasNoFileExtension = /\/[^\/\.]*(\?|$)/;
+
+/**
+ * Returns a promise that resolves with a generated service worker
+ * configuration.
+ */
+export async function generateServiceWorkerConfig(
+    options: AddWorkboxServiceWorkerOptions): Promise<WorkboxConfig> {
+  assert(!!options, '`project` & `buildRoot` options are required');
+  assert(!!options.project, '`project` option is required');
+  assert(!!options.buildRoot, '`buildRoot` option is required');
+
+  options = {...options};
+  const project = options.project;
+  const buildRoot = options.buildRoot;
+  const workboxConfig: WorkboxConfig = {
+    ...options.workboxConfig,
+    globDirectory: buildRoot,
+    navigateFallback: path.relative(project.config.root, project.config.entrypoint),
+  };
+
+  const depsIndex = await project.analyzer.analyzeDependencies;
+  let staticFileGlobs = Array.from(workboxConfig.globPatterns || []);
+  const precachedAssets = (options.bundled) ?
+      getBundledPrecachedAssets(project) :
+      getPrecachedAssets(depsIndex, project);
+
+  if (workboxConfig.globPatterns === undefined) {
+    staticFileGlobs = staticFileGlobs.concat(precachedAssets);
+    staticFileGlobs = staticFileGlobs.map((filePath: string) => {
+      if (filePath.startsWith(project.config.root)) {
+        filePath = filePath.substring(project.config.root.length);
+      }
+      return removeLeadingSlash(filePath);
```

Comment:
I'm taking a look at the Windows CI test failure at https://ci.appveyor.com/project/Polymer/tools/build/1.0.1475#L1409 and one thing that stands out is that the string `'\\shell.html'` is being used as a glob pattern. That shouldn't be what's used, since glob patterns always use `/` as a path separator, even on windows.

I feel like the `staticFileGlobs.concat(precachedAssets)` line a few lines up might be introducing paths with `'\\'` characters into the list of glob patterns, and then this line of code isn't sufficient to remove that leading `'\\'`.

But in general, all glob patterns need to use `'/'`, regardless of where the separator occurs.

Reply:
Thanks for the suggestion @jeffposnick , fixed it